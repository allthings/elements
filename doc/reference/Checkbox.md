<!-- 
This is an auto-generated markdown. 
You can change it in "/Users/daniel/Dev/allthings/elements/src/Checkbox/Checkbox.tsx" and run build:docs to update this file.
-->
# Checkbox
Checkbox are used to give users a way to select or deselect options.

```example
<View>
   <Checkbox checked name="ok" label="Are you ok?" />
   <Checkbox name="notok" label="Are you not ok?" />
</View>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|checked|boolean|True to make it checked
|name **(required)**||Label of Checkbox
|onChange **(required)**||Called, when the user changes something
|defaultValue|string, string[]|The default value to put into the component, without making it controlled
|required|boolean|Indicates that this field is required
|icon|"download", "alarm", "alarm-filled", "armchair", "armchair-filled", "arrow-down", "arrow-down-filled", "arrow-left", "arrow-left-filled", "arrow-right", "arrow-right-filled", ... 146 more ...|Icon shown on the left of the input field (See `Icon` component for all possible values) *
|label|string|The label of the input
|type|string|Type, can be: 'tel', 'number', 'text', 'url', 'email'
|value|string, number, string[]|The value, makes this component a controlled component
|lines|number|Can only be used with type=text. Increase to enable multi-line input
|hasRightIcon|boolean|Used when there is an icon to the right of input field
|pattern|string|Regular expression to validate against
|minLength|number|Min number of characters that must be provided
|maxLength|number|Max number of characters that can be provided
|onInputRef|any|Called with the input field a reference
|readOnly|boolean|
|disabled|boolean|
|forwardedRef|any|
|badInput|string|Error message for bad input
|customError|string|Error message for customError
|patternMismatch|string|Error message for patternMismatch
|rangeOverflow|string|Error message for rangeOverflow
|rangeUnderflow|string|Error message for rangeUnderflow
|stepMismatch|string|Error message for stepMismatch
|tooLong|string|Error message for tooLong
|tooShort|string|Error message for tooShort
|typeMismatch|string|Error message for typeMismatch
|valueMissing|string|Error message for valueMissing
