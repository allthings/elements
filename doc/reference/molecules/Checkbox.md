<!-- 
This is an auto-generated markdown. 
You can change it in "src/molecules/Checkbox.jsx" and run build:docs to update this file.
-->
# Checkbox
Checkbox are used to give users a way to select or deselect options.

```example
<View>
  <Checkbox checked name="ok" label="Are you ok?" />
  <Checkbox name="notok" label="Are you not ok?" />
</View>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|checked|bool|True to make it checked
|label **(required)**|union|Label of Checkbox
|labelSize|custom|Text size of the label<br>Default: 'l'
|name **(required)**|string|
|onChange|func|<br>Default: () => {}
|backgroundColor|string|Background color of the form item
|hideLine|bool|ListItem props<br>Default: false
|padded|bool|<br>Default: true
