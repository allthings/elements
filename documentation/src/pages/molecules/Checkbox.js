import React from 'react'
import { Checkbox, View } from '@allthings/elements'
import Example from '../../components/Example'
import Notes from '../../components/Notes'

export default () => (
  <View>
    <Notes for={Checkbox} />
    <Example>
      {`<Checkbox checked name="ok" label="Are you ok?" />
<Checkbox name="notok" label="Are you not ok?" />`}
    </Example>
  </View>
)
