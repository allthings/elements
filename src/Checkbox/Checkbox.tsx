import { css } from 'glamor'
import React, { ChangeEvent, useState } from 'react'
import Absolute from '../Absolute'
import Icon from '../Icon'
import { IInputProps } from '../Input'
import Relative from '../Relative'
import { useTheme } from '../Theme'
import View from '../View'

const styles = {
  checkbox: (background: string, checked: boolean) =>
    css({
      borderRadius: '3px',
      height: '25px',
      width: '25px',
      backgroundColor: checked && background,
      borderWidth: 3,
      borderStyle: 'solid',
      borderColor: checked ? background : 'lightGrey',
      transition: '250ms',
      cursor: 'pointer',
    }),
}
export interface ICheckboxProps extends IInputProps {
  /** True to make it checked */
  readonly checked?: boolean
  /** Label of Checkbox */
  readonly name?: string
  readonly onChange?: (e: ChangeEvent<HTMLInputElement>) => void
}

/**
 * Checkbox are used to give users a way to select or deselect options.
 *
 * ```example
 * <View>
 *   <Checkbox checked name="ok" label="Are you ok?" />
 *   <Checkbox name="notok" label="Are you not ok?" />
 * </View>
 * ```
 */
const Checkbox = ({ checked, name, onChange, ...props }: ICheckboxProps) => {
  const isControlled = typeof checked !== 'undefined'

  const [isChecked, setIsChecked] = useState(checked)

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (!isControlled) {
      setIsChecked(checked => !checked)
    }
    onChange && onChange(e)
  }

  const checkedValue = Boolean(isControlled ? checked : isChecked)
  const { colorize } = useTheme()

  return (
    <View direction="row" alignV="center">
      <Relative
        direction="row"
        alignV="center"
        alignH="center"
        {...styles.checkbox(colorize('primary'), checkedValue)}
      >
        {checkedValue && (
          <Relative bottom={1}>
            <Icon name="check-filled" size={14} color="#fff" />
          </Relative>
        )}
        <Absolute top={0} left={0} right={0} bottom={0}>
          <input
            type="checkbox"
            checked={checkedValue}
            id={name}
            name={name}
            value={String(checkedValue)}
            style={{
              opacity: 0,
              width: '25px',
              height: '25px',
              margin: 0,
              cursor: 'pointer',
            }}
            onChange={handleChange}
            {...props}
          />
        </Absolute>
      </Relative>
    </View>
  )
}

export default Checkbox
