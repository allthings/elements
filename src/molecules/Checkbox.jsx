import React from 'react'
import View from '../atoms/View'
import Icon from '../atoms/Icon'
import Text from '../atoms/Text'
import Relative from '../atoms/Relative'
import ListItem from './List/ListItem'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'
import PropTypes from 'prop-types'
import Absolute from '../atoms/Absolute'
import Inset from '../atoms/Inset'

const styles = {
  checkbox: (background, checked) =>
    css({
      borderRadius: '3px',
      height: '25px',
      width: '25px',
      backgroundColor: checked && background,
      borderWidth: 3,
      borderStyle: 'solid',
      borderColor: checked ? background : 'lightGrey',
      transition: '250ms',
    }),
  text: css({
    width: 200,
    marginLeft: 20,
  }),
}

/**
 * Checkbox are used to give users a way to select or deselect options.
 *
 * ```example
 * <View>
 *   <Checkbox checked name="ok" label="Are you ok?" />
 *   <Checkbox name="notok" label="Are you not ok?" />
 * </View>
 * ```
 */
class Checkbox extends React.Component {
  static propTypes = {
    /** True to make it checked */
    checked: PropTypes.bool,
    /** Label of Checkbox */
    label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
    /** Text size of the label */
    labelSize: Text.propTypes.size,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    /** Background color of the form item */
    backgroundColor: PropTypes.string,
  }

  state = {
    checked: this.props.checked,
  }

  static defaultProps = {
    checked: false,
    labelSize: 'l',
  }

  handleChange = () => this.setState(({ checked }) => ({ checked: !checked }))

  render() {
    const {
      checked,
      onChange,
      label,
      labelSize,
      name,
      backgroundColor,
      ...props
    } = this.props
    const realChecked = checked || this.state.checked
    const changeHandler = onChange || this.handleChange
    return (
      <Theme>
        {({ theme, colorize }) => (
          <ListItem backgroundColor={colorize(backgroundColor)}>
            <View direction="row" alignV="center">
              <Relative
                direction="row"
                alignV="center"
                alignH="center"
                {...styles.checkbox(theme.primary, realChecked)}
              >
                {realChecked && (
                  <Relative bottom={1}>
                    <Icon name="check-filled" size={14} color="#fff" />
                  </Relative>
                )}
                <Absolute top={0} left={0} right={0} bottom={0}>
                  <input
                    type="checkbox"
                    checked={realChecked}
                    id={name}
                    value={realChecked}
                    style={{
                      opacity: 0,
                      width: '25px',
                      height: '25px',
                      margin: 0,
                    }}
                    onChange={changeHandler}
                    {...props}
                  />
                </Absolute>
              </Relative>
            </View>
            <label htmlFor={name}>
              <Inset horizontal>
                <Text size={labelSize}>{label}</Text>
              </Inset>
            </label>
          </ListItem>
        )}
      </Theme>
    )
  }
}

export default Checkbox
