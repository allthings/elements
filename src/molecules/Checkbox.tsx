import React, { Component, ChangeEvent } from 'react'
import View from '../atoms/View'
import Icon from '../atoms/Icon'
import Relative from '../atoms/Relative'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'
import Absolute from '../atoms/Absolute'
import { IInputProps } from '../atoms/Input'

const styles = {
  checkbox: (background: string, checked: boolean) =>
    css({
      borderRadius: '3px',
      height: '25px',
      width: '25px',
      backgroundColor: checked && background,
      borderWidth: 3,
      borderStyle: 'solid',
      borderColor: checked ? background : 'lightGrey',
      transition: '250ms',
      cursor: 'pointer',
    }),
}
export interface ICheckboxProps extends IInputProps {
  /** True to make it checked */
  checked?: boolean
  /** Label of Checkbox */
  name: string
  onChange: (e: ChangeEvent) => void
}
interface IState {
  checked?: boolean
}
/**
 * Checkbox are used to give users a way to select or deselect options.
 *
 * ```example
 * <View>
 *   <Checkbox checked name="ok" label="Are you ok?" />
 *   <Checkbox name="notok" label="Are you not ok?" />
 * </View>
 * ```
 */
class Checkbox extends Component<ICheckboxProps, IState> {
  state = {
    checked: this.props.checked,
  }

  isControlled = () => typeof this.props.checked !== 'undefined'

  handleChange = (e: ChangeEvent) => {
    if (!this.isControlled()) {
      this.setState(({ checked }) => ({ checked: !checked }))
    }
    this.props.onChange && this.props.onChange(e)
  }

  render() {
    const { checked, name, onChange, ...props } = this.props
    const realChecked = Boolean(
      this.isControlled() ? checked : this.state.checked
    )

    return (
      <Theme>
        {({ colorize }) => (
          <View direction="row" alignV="center">
            <Relative
              direction="row"
              alignV="center"
              alignH="center"
              {...styles.checkbox(colorize('primary'), realChecked)}
            >
              {realChecked && (
                <Relative bottom={1}>
                  <Icon name="check-filled" size={14} color="#fff" />
                </Relative>
              )}
              <Absolute top={0} left={0} right={0} bottom={0}>
                <input
                  type="checkbox"
                  checked={realChecked}
                  id={name}
                  name={name}
                  value={String(realChecked)}
                  style={{
                    opacity: 0,
                    width: '25px',
                    height: '25px',
                    margin: 0,
                    cursor: 'pointer',
                  }}
                  onChange={this.handleChange}
                  {...props}
                />
              </Absolute>
            </Relative>
          </View>
        )}
      </Theme>
    )
  }
}

export default Checkbox
