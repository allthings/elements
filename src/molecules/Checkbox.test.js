import React from 'react'
import Checkbox from './Checkbox'
import { ThemeProvider } from '../'

const THEME = {
  background: 'darkMarco',
}
const STRING_LABEL = 'I am a nice checkbox!'

describe('Test the checkbox component', () => {
  it('should tick it', () => {
    const wrapper = mount(
      <ThemeProvider theme={THEME}>
        <Checkbox label={STRING_LABEL} name="a" />
      </ThemeProvider>
    )
    const input = wrapper.find('input#a')
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('label').text()).toEqual(STRING_LABEL)
    expect(input.prop('checked')).toBeFalsy()
    // Check it.
    input.simulate('change')
    // Re-render in order to get it checked.
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('input#a').prop('checked')).toBeTruthy()
    wrapper.unmount()
  })
  it('should work with children', () => {
    const wrapper = mount(
      <ThemeProvider theme={THEME}>
        <Checkbox label={<h1>test</h1>} name="a" backgroundColor="darkMarco" />
      </ThemeProvider>
    )
    expect(wrapper).toMatchSnapshot()
    wrapper.unmount()
  })
})
