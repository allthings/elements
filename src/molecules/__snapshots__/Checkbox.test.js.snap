// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the checkbox component should tick it 1`] = `
.css-0,
[data-css-0] {
  position: relative;
}

.css-1,
[data-css-1] {
  border-radius: 3px;
  height: 25px;
  width: 25px;
  border-width: 3px;
  border-style: solid;
  border-color: lightGrey;
  transition: 250ms;
  -webkit-transition: 250ms;
  -moz-transition: 250ms;
}

.css-2,
[data-css-2] {
  padding: 10px 15px;
  min-height: 50px;
  border-bottom: 1px solid #ecf0f1;
  background-color: white;
}

.css-3,
[data-css-3] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-4,
[data-css-4] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

.css-5,
[data-css-5] {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

.css-6,
[data-css-6] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-7,
[data-css-7] {
  padding-left: 15px;
  padding-right: 15px;
}

.css-8,
[data-css-8] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

<div
  data-css-2=""
  data-css-3=""
>
  <div
    data-css-3=""
  >
    <div
      data-css-0=""
      data-css-1=""
      data-css-4=""
    >
      <div
        data-css-6=""
        data-css-5=""
      >
        <input
          checked={false}
          id="a"
          name="a"
          onChange={[Function]}
          style={
            Object {
              "height": "25px",
              "margin": 0,
              "opacity": 0,
              "width": "25px",
            }
          }
          type="checkbox"
          value={false}
        />
      </div>
    </div>
  </div>
  <label
    htmlFor="a"
  >
    <div
      data-css-6=""
      data-css-7=""
    >
      <div
        data-css-8=""
        data-css-6=""
      >
        I am a nice checkbox!
      </div>
    </div>
  </label>
</div>
`;

exports[`Test the checkbox component should tick it 2`] = `
.css-0,
[data-css-0] {
  position: relative;
}

.css-1,
[data-css-1] {
  padding: 10px 15px;
  min-height: 50px;
  border-bottom: 1px solid #ecf0f1;
  background-color: white;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

.css-4,
[data-css-4] {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

.css-5,
[data-css-5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-6,
[data-css-6] {
  padding-left: 15px;
  padding-right: 15px;
}

.css-7,
[data-css-7] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-8,
[data-css-8] {
  border-radius: 3px;
  height: 25px;
  width: 25px;
  background-color: #3598db;
  border-width: 3px;
  border-style: solid;
  border-color: #3598db;
  transition: 250ms;
  -webkit-transition: 250ms;
  -moz-transition: 250ms;
}

.css-9,
[data-css-9] {
  position: relative;
  bottom: 1px;
}

<div
  data-css-1=""
  data-css-2=""
>
  <div
    data-css-2=""
  >
    <div
      data-css-0=""
      data-css-3=""
      data-css-8=""
    >
      <div
        data-css-9=""
        data-css-5=""
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          data-css-5=""
          style={
            Object {
              "fill": "#fff",
              "height": 14,
              "stroke": false,
              "width": 14,
            }
          }
        />
      </div>
      <div
        data-css-5=""
        data-css-4=""
      >
        <input
          checked={true}
          id="a"
          name="a"
          onChange={[Function]}
          style={
            Object {
              "height": "25px",
              "margin": 0,
              "opacity": 0,
              "width": "25px",
            }
          }
          type="checkbox"
          value={true}
        />
      </div>
    </div>
  </div>
  <label
    htmlFor="a"
  >
    <div
      data-css-5=""
      data-css-6=""
    >
      <div
        data-css-7=""
        data-css-5=""
      >
        I am a nice checkbox!
      </div>
    </div>
  </label>
</div>
`;

exports[`Test the checkbox component should work with children 1`] = `
.css-0,
[data-css-0] {
  position: relative;
}

.css-1,
[data-css-1] {
  border-radius: 3px;
  height: 25px;
  width: 25px;
  border-width: 3px;
  border-style: solid;
  border-color: lightGrey;
  transition: 250ms;
  -webkit-transition: 250ms;
  -moz-transition: 250ms;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

.css-4,
[data-css-4] {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

.css-5,
[data-css-5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-6,
[data-css-6] {
  padding-left: 15px;
  padding-right: 15px;
}

.css-7,
[data-css-7] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-8,
[data-css-8] {
  padding: 10px 15px;
  min-height: 50px;
  border-bottom: 1px solid #ecf0f1;
  background-color: darkMarco;
}

<div
  data-css-2=""
  data-css-8=""
>
  <div
    data-css-2=""
  >
    <div
      data-css-0=""
      data-css-1=""
      data-css-3=""
    >
      <div
        data-css-5=""
        data-css-4=""
      >
        <input
          checked={false}
          id="a"
          name="a"
          onChange={[Function]}
          style={
            Object {
              "height": "25px",
              "margin": 0,
              "opacity": 0,
              "width": "25px",
            }
          }
          type="checkbox"
          value={false}
        />
      </div>
    </div>
  </div>
  <label
    htmlFor="a"
  >
    <div
      data-css-5=""
      data-css-6=""
    >
      <div
        data-css-7=""
        data-css-5=""
      >
        <h1>
          test
        </h1>
      </div>
    </div>
  </label>
</div>
`;
